### ISSUE TITLE
Implementation of Spatial Prior

### ISSUE THREAD

**Issue by ShigemichiMatsuzaki** [yzou2/CRST#15]
Hi,
Thank you for your excellent work.

I'm currently trying to implement spatial prior(SP) as presented in your CBST paper (at this moment, only in the pretraining), since it's apparently not implemented in this pytorch version.
After applying the SP, I guess the resulting output scores will be a few orders of magnitude less than those without applying the SP, which results in so small gradient that the training doesn't proceed well.

I've come up with several measures for this:
1. Simply multiplying the resulting output by a random number (like, 10e+4) before calculating cross entropy loss.
2. Normalize the output values of a pixel over the classes (via softmax or simply dividing by the sum)
3. Increase the learning rate.

I looked it up on google, but couldn't find exact answer for my question.
Could you give me an advice on it?
Sorry for the silly question.

Thanks for your work again, and I'm happy to hear from you.

**Comment by yzou2** [comment]
Thanks a lot for your interests on our works. You are right about the magnitude of output score after multiplying the SP. But the SP is used to generate one-hot/soft pseudo-labels only. Thus the magnitude order does not matter. In the training, if you look at the equation (9) in our CBST paper (https://arxiv.org/pdf/1810.07911.pdf), y_hat * log (q * p) = y_hat * log q + y_hat * log p where y_hat * log q is constant. The gradient produced by the loss only flows to the  y_hat * log p.

**Comment by ShigemichiMatsuzaki** [comment]
Hi,
Thank you for your quick and kind reply.

I totally misunderstood the equation.
I'll review the paper again.

Thanks a lot!
